Add WhatsApp importer

jphastings opened the pull request:
A first pass at importing WhatsApp chat exports.

As a first time contributor to an early-stage project I'd very much appreciate strong steer if this is in the right ballpark!
I've managed to import multi-GB chats of my own without problem with this code — but I'm probably missing something important in the wider context of the project!

## Context

Some high level context on the WhatsApp exports
- They're _very_ low fidelity exports (see [my example](https://github.com/jphastings/timelinize/blob/whatsapp/datasources/whatsapp/testdata/fixtures/_chat.txt)) — they omit lots of data that would be useful in a backup (see below in "annoying quirks")
- You can [export a chat](https://faq.whatsapp.com/1180414079177245/) to test with this by tapping "Export chat" at the bottom of your chat settings (where you'd add new people to the chat)

## Open questions

- ~~**Do we want to import context messages?** ("you deleted this message", and "Messages and calls are end-to-end encrypted. Only people in this chat can read, listen to, or share them." at the top of every chat)~~
    - We've agreed these aren't necessary
- In WhatsApp its possible to have groups with the same participants but a different group name. **Is it possible to tag a conversation with a "group name" in Timelinize?** If not, this may end up with different conversations being interleaved within Timelinize.
    - We're discussing in #81
- ~~**Is it safe to assume the current location for timezone analysis of imports?** WhatsApp exports use timezoneless timestamps (ugh). For my exports (and I believe all exports) the timestamps are in the timezone of the phone at the time of export, so messages sent while abroad will _look_ to be sent at odd hours of the day, but will be accurate _if_ the timezone of Timelinize at time of import was the same as the timezone of the phone on export.~~
    - This is reasonable, the docs & UI will make this clear.
- ~~**Is there a way to tag images as stickers in Timelinize?** I _think_ I can reliably detect when an attached image is a sticker, it might be nice to have them be flagged as such in Timelinize, otherwise you effectively get a low quality image that shows up very large on the timeline (as they're often square).~~
    - Being discussed in #80
- ~~Image/attachment posts currently _will never_ have caption text (see annoying quirk below), but I've still modelled them as a root `Item` without any Content, and a `RelAttachment` `Item` with the attached file's data. **Is data modelling like this the right choice?** (I'm considering the case where, if Meta fixes this issue, we'd probably want a subsequent import to add the text to the root `Item`, and leave the attachment as it is, rather than replacing the data root `Item` with a text root `Item` and adding a new attachment with the same data as was replaced in the root `Item`)~~
    - We've agreed this is a good way to model the messages. I've also added a `Retrieve` key so we can add a caption, if they arrive in the future.
- ~~**Do we want to do something special with Polls?** (WhatsApp allows polling people for multiple choice answers. The [exports](https://github.com/jphastings/timelinize/blob/whatsapp/datasources/whatsapp/testdata/fixtures/_chat.txt#L13-L16) don't hold who voted what, but they do hold total votes for each option)~~
    - Polls are now parsed (in multiple languages) & their metadata extracted into `Poll *` metadata keys.
- ~~**Do we want to do something special with sent locations?** The exports either show a [Name & Foursquare URL](https://github.com/jphastings/timelinize/blob/whatsapp/datasources/whatsapp/testdata/fixtures/_chat.txt#L17) (for when a named location is selected) or a [Google Maps link](https://github.com/jphastings/timelinize/blob/whatsapp/datasources/whatsapp/testdata/fixtures/_chat.txt#L18) (with an extractable lat/long, for when "my current location" is sent).~~
    - Locations are now extracted into the `Location *` metadata keys.
- ~~I _think_ the attached Image/Video/file `Item`s should be `Classification: timeline.ClassMessage`, as they are now — this is what the Facebook importer seems to do. **Am I right in assuming this?**~~
    - We've agreed this is suitable.
- WhatsApp allows for some (slightly uncommon) text formatting within its messages. (`_italic_`, `*bold*` (different to markdown), backticks for monospaced, `~strikethrough~` (non-standard markdown)) — given these are likely to be different across platforms, **do you want to store formatting in a Timelinize-specific manner?**
    - I'll address this in a subsequent PR.

## Quirks

These are the annoying quirks of the export format we should find good ways to communicate to people using Timelinize:

- Any caption text sent with an attachment isn't exported by WhatsApp. (The text is lost and unavailable to Timelinize — I've opened a bug with Meta, for all the good that'll do)
- If there are silent members of a group chat, their presence isn't recorded in the data WhatsApp exports
- Related: Because of this, we have to hold all messages in memory while processing the entire file before emitting them (so we can correctly add all recipients, even if one spoke for the first time on the last message), so very large chat logs may use a lot of memory & file descriptors (if there are many attachments)
- People using Timelinize will need to export _each conversation manually_ from WhatsApp. Which is a pain. You can't export from WhatsApp web (which would be easy to script, with an extension), so automating this will be tricky/impossible.
- Reactions, events, call indicators (and obviously the actual phonecalls), and likely other fancy parts of the conversation, just aren't exported.


## Asides:

These are things I stumbed across that might be useful as broader features/bugs for timelinize — let me know if you'd like me to record them in more detail somewhere else, or try to fix them in a different PR (I know Timelinize is in an early phase of life!)

- The photos shown in the "conversations" section of the webapp are cropped, but I struggled to find a way to embiggen them (I expected to click on them to have them expand, finally found the "…" → "View Item")
  - Related: using "View item" to look at an image in detail loses my place in the timeline when I press back in my browser.
  - Related: multiple consecutive images are (very nicely!) grouped together in Timelinize's chat, but clicking the "View Item" button on "…" only takes me to the first, so there's no way to embiggen the 2nd or onward photos.
- I happened to have some PDF and ZIP files sent in one of my imports. They currently show just `Unsupported format (application/pdf)`, it'd be great if that was a link, so I could download them/fall back on my browser's capabilities.
- ~~Some WhatsApp voice notes are in Ogg Vorbis containers, which don't seem to be supported by Timelinize yet.~~ Fixed!

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/timelinize/timelinize?pullRequest=79) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/timelinize/timelinize?pullRequest=79) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/timelinize/timelinize?pullRequest=79) it.</sub>

mholt commented:
Hey JP! Thanks for popping into our Discord. I'm excited to review and merge this!

Still on toddler duty, but let me reply real quick to a few things while I have a chance:

> Do we want to import context messages? ("you deleted this message", and "Messages and calls are end-to-end encrypted. Only people in this chat can read, listen to, or share them." at the top of every chat)

Probably not? Maybe if you think it's useful to follow a conversation. Otherwise, probably not really meaningful in an archival setting.

> In WhatsApp its possible to have groups with the same participants but a different group name. Is it possible to tag a conversation with a "group name" in Timelinize? If not, this may end up with different conversations being interleaved within Timelinize.

Hmm, good question. In Timelinize, conversations are _implied_. They're discovered by querying for message-class items that have "sent" relations to the same group of entities/people.

However, there are also "collection" relations, like for adding photos or music to albums. The UI doesn't currently show collections, but the DB supports it. That might be a good way to group the messages and then in the future you could group by collection relations.

There's also the concept of tags in Timelinize schema; but I haven't developed this out yet. This is worthy of more discussion! (Open its own issue?)

> Is it safe to assume the current location for timezone analysis of imports? WhatsApp exports use timezoneless timestamps (ugh). For my exports (and I believe all exports) the timestamps are in the timezone of the phone at the time of export, so messages sent while abroad will look to be sent at odd hours of the day, but will be accurate if the timezone of Timelinize at time of import was the same as the timezone of the phone on export.

Yeah, I hate it when data sources don't give proper timestamps (with time zones). In this case I think just assuming the current local time is fine... is it "safe"? No :smile: But I don't know what else we can do. With other data sources, we can possibly cross-correlate geolocation for the same person and fix time zones. (Actually, that'd be an _awesome_ and unique feature for Timelinize: after importing your location history or photo library, push a button to fix up timestamps of items without a time zone!)

> Is there a way to tag images as stickers in Timelinize? I think I can reliably detect when an attached image is a sticker, it might be nice to have them be flagged as such in Timelinize, otherwise you effectively get a low quality image that shows up very large on the timeline (as they're often square).

We kind of have "item classifications" for that. Right now, all images are just "Media". I've thought about differentiating "Screenshots" and "Scans" and traditional "Media" (photo/video library). Maybe "Sticker" could be another class... the Facebook "like" image is another example, but in that case I don't think we know if it's the "like" image or not.

The schema DOES support the concept of tags, but as I said, I haven't developed it out yet. Let's discuss this more. Maybe a separate issue, if you want to start that?

> Image/attachment posts currently will never have caption text (see annoying quirk below), but I've still modelled them as a root Item without any Content, and a RelAttachment Item with the attached file's data. Is data modelling like this the right choice? (I'm considering the case where, if Meta fixes this issue, we'd probably want a subsequent import to add the text to the root Item, and leave the attachment as it is, rather than replacing the data root Item with a text root Item and adding a new attachment with the same data as was replaced in the root Item)

That's probably how I would do it, too. If you can set a RetrievalKey on the empty item in a reliable way (think of it like a unique ID, but it's something you made up as opposed to the data source giving it to you -- I assume they don't give you IDs), then a future import of a "fixed" data set could "fill out" the item the rest of the way.

> Do we want to do something special with Polls? (WhatsApp allows polling people for multiple choice answers. The [exports](https://github.com/jphastings/timelinize/blob/whatsapp/datasources/whatsapp/testdata/fixtures/_chat.txt#L13-L16) don't hold who voted what, but they do hold total votes for each option)

I don't know if I have anything in mind for them. Could probably just be an item with preformatted content I guess? Or maybe metadata to slightly give some structure to the poll results?

> Do we want to do something special with sent locations? The exports either show a [Name & Foursquare URL](https://github.com/jphastings/timelinize/blob/whatsapp/datasources/whatsapp/testdata/fixtures/_chat.txt#L17) (for when a named location is selected) or a [Google Maps link](https://github.com/jphastings/timelinize/blob/whatsapp/datasources/whatsapp/testdata/fixtures/_chat.txt#L18) (with an extractable lat/long, for when "my current location" is sent).
This could probably be handled later by whatever Timelinize decides to do with URLs in messages/posts.

Hmmm, probably not? (Other than special URL handling, which I haven't implemented yet.)

> I think the attached Image/Video/file Items should be Classification: timeline.ClassMessage, as they are now — this is what the Facebook importer seems to do. Am I right in assuming this?

Yes, since they were sent as a message. Good question! ClassMedia is for standalone images/videos/etc.

> WhatsApp allows for some (slightly uncommon) text formatting within its messages. (_italic_, *bold* (different to markdown), backticks for monospaced, ~strikethrough~ (non-standard markdown)) — given these are likely to be different across platforms, do you want to store formatting in a Timelinize-specific manner?

I don't have a special formatting standard... like the content could be text/plain, or text/markdown, or text/html, for example, and it's up to the UI to render it accordingly. So you could choose to render formatted messages as HTML or Markdown if you can convert it? Just set the MediaType accordingly.

---

Will reply more later!

jphastings commented:
Nice! I'll fixup the code/create new issues & fix the linting fail then post back, probably later today or tomorow — thanks for the rapid response (particulalry with a toddler to look after! 😄)

mholt commented:
Quick follow-ups:

- The "Quirks" section is all great to note in our docs on our website. I'll open source that soon.

> The photos shown in the "conversations" section of the webapp are cropped, but I struggled to find a way to embiggen them (I expected to click on them to have them expand, finally found the "…" → "View Item")

Yeah, that UI definitely needs some work. If nothing else they should be links. I'm also experimenting with a modal popup like in the gallery view.

> Related: using "View item" to look at an image in detail loses my place in the timeline when I press back in my browser.

Yep, this sucks, and is why I'm inclined to use a modal popup to preview an item in-situ.

> Related: multiple consecutive images are (very nicely!) grouped together in Timelinize's chat, but clicking the "View Item" button on "…" only takes me to the first, so there's no way to embiggen the 2nd or onward photos.

The Item page should, ideally, show ALL information and related items. We probably just aren't doing that all in the UI yet.

> I happened to have some PDF and ZIP files sent in one of my imports. They currently show just Unsupported format (application/pdf), it'd be great if that was a link, so I could download them/fall back on my browser's capabilities.

Agreed, PDF support is planned actually.

> Some WhatsApp voice notes are in Ogg Vorbis containers, which don't seem to be supported by Timelinize yet.

Ohh, yeah can you tell me more about this?

jphastings commented:
Thanks for all the notes! I'm working on code changes & other things, but to quickly get back to you on the Ogg Vorbis point:

Here's a zip of a sample file I have direct from a WhatsApp export: [00000022-AUDIO-2025-05-03-17-40-28.opus.zip](https://github.com/user-attachments/files/20024183/00000022-AUDIO-2025-05-03-17-40-28.opus.zip)

```
ffprobe ~/Desktop/00000022-AUDIO-2025-05-03-17-40-28.opus
ffprobe version 7.1.1 Copyright (c) 2007-2025 the FFmpeg developers
( snip )
Input #0, ogg, from '00000022-AUDIO-2025-05-03-17-40-28.opus':
  Duration: 00:00:02.53, start: 0.006500, bitrate: 17 kb/s
  Stream #0:0: Audio: opus, 48000 Hz, mono, fltp
```

It looks like this on my imports:

<img width="860" alt="Screenshot 2025-05-03 at 17 43 13" src="https://github.com/user-attachments/assets/bcd90765-9934-4984-a941-8190a9f6df0a" />

mholt commented:
Thanks! I wasn't familiar with the .opus extension, nor the fact that it's "application/ogg" instead of "audio/ogg"... so the UI needs to check for that special case I guess.

mholt commented:
@jphastings The latest commit might play better with that file, I haven't tried it but just a quick patch that seemed correct either way.

jphastings commented:
Awesome — looks like your quick fix sorted out the opus audio snippets in my exports!

<img width="852" alt="Screenshot 2025-05-03 at 23 15 31" src="https://github.com/user-attachments/assets/b3e3806b-9e67-4b26-92c2-6c331bf024e1" />

jphastings commented:
I've added a couple of commits in which solve a few problems (and possibly create one or two more 😅):

- I've added Retrieval keys to all the messages as you suggested — I'm using "whatsapp", the timestamp and the name of the sender as the best option; if people rename one of their contacts then there will be duplicate messages, but I think this is unavoidable.
- Depending on the locale settings of your phone when you export, the date stamps may be in a number of different formats — my last commit handles almost all automatically, but will assume DD/MM/YYYY rather than MM/DD/YYYY (we may need to add an import option to flip to/from American format?)
- Location data is pulled out as metadata (the `Location Latitude` & `Location Longitude`, or `Location Foursquare ID`)
- Poll data is de-localised for the message (just uses markdown-style bullets & symbols to represent options and votes), and  also extracted as metadata (`Poll Question`, `Poll Option N`, `Poll Votes N`, where N is 1, 2, 3… (1 indexed) depending on the number of options)

⚠️ There's also a bug I've just spotted with the test data being incorrect for polls (`OPTION:` lines should be prefixed with the `\u200E` character) — which I'll fix when I next have a spare moment 😊

jphastings commented:
✅ Fixed that little bug; Polls aren't pretty to look at, but the data's all there 😄

Are these, and the above metadata names, the right ones to be using? 

<img width="859" alt="Screenshot 2025-05-03 at 23 51 08" src="https://github.com/user-attachments/assets/44807ec8-8f9f-4241-8cfb-d95793e2991b" />

jphastings commented:
Right; a barrage of messages and commits today — sorry for the flurry! To wrap up the day, I believe I've addressed all the points we discussed in code except for:

- Creating a collection of all the messages from the same import (as a "group chat") — I'll leave that for another PR, as I think it'll be a complex one
- I opened #80 to consider annotating stickers
- I've not yet parsed WhatsApp's formatting — I'll leave this for another PR as well (I'll parse into HTML, as you suggest, which will mean escaping existing tags & building a parser that matches WhatsApp's format as closely as possible — this will non-trivial 😅)

…which I think covers everything! I've not been able to get the linter working on my machine, so there may be some grumbles there, which I'll be happy to fix if they crop up. Do shout if there's anything you think needs altering, as and when you get the chance to go deeper.

Thanks again for putting Timelinize together, and being so welcoming to me & for my contributions today!

mholt commented on datasources/whatsapp/whatsapp.go:
```diff
@@ -0,0 +1,183 @@
+			Name: name,
+			Attributes: []timeline.Attribute{
+				{
+					Name:     "WhatsApp Name",
```
```suggestion
					Name:     "whatsapp_name",
```

I should clarify in the docs that attribute names are snake_cased... they're mostly internal and act kind of like keys in the DB; though the UI can have a mapping to human-friendly names for common attributes.

jphastings replied:
Ahh! I started with the snake_case, but saw it come through as snake in the UI, so swapped it 😅 Thanks!

mholt commented on datasources/whatsapp/whatsapp.go:
```diff
@@ -0,0 +1,183 @@
+// Recognize returns whether the input is supported.
+func (Importer) Recognize(_ context.Context, dirEntry timeline.DirEntry, _ timeline.RecognizeParams) (timeline.Recognition, error) {
+	if dirEntry.FileExists(chatPath) {
+		return timeline.Recognition{Confidence: 1}, nil
```
That's pretty high confidence for just a single filename check; maybe a slightly lower value? As opposed to actually opening the file and verifying the contents, which could be more like 1. I'd say this could be more like, I dunno, .8?

jphastings replied:
Very happy to take your steer here — I wasn't quite sure what the confidence rating would be relative to! 😅 

(For what it's worth I figured "other references to `_chat.txt` as a file name" would be a good proxy to the score, did a [web search](https://kagi.com/search?q=%22_chat.txt%22+-whatsapp&r=gb&sh=q0FkCQFV74NomUfNSaD1rA), found _only_ references to WhatsApp, and assumed only maliciously crafted files would ever be a match and _not_ a WhatsApp export.)

mholt reviewed (changes requested):
Thanks for the great contribution! As I don't use WhatsApp I haven't verified this myself, but the Timelinize-related code mostly LGTM from a quick scan. Just a nit regarding an attribute name so far.

mholt commented:
The poll metadata looks good too -- though, in the future, we may want to tweak it a bit based on experience.

I'll try to fix the CSS for the poll rendering later as well.

mholt commented on datasources/whatsapp/locations.go:
```diff
@@ -0,0 +1,38 @@
+			return "", nil, false
+		}
+
+		meta["Location Latitude"] = lat
+		meta["Location Longitude"] = lng
```
```suggestion
		meta["Latitude"] = lat
		meta["Longitude"] = lng
```

How would you feel about this?

It's a bit unconventional, I suppose, since normally you'd put this in the item's geocoords directly; but that would imply the item was at that location. Is that the case of these messages? In one sense, it is, since the message is about that location, even if it is not describing a person being at that location. So, I'm open to discussing this now that I see it.

jphastings replied:
I was wondering about exactly this!

I think it'd be worth being clear in your data model that this location is "over there", rather than "over here". (These are sometimes a 'current location', but the likelihood is vanishingly small, as the current location messages are auto deleted after their fixed timespan of ~1 hour — and few people will be taking an export in that hour.) I'd worry that putting them without a namespace might make them easier to misunderstand them as being about "where the sender is"?

The second reason I went with a namespace is the "named location" pins are Foursquare IDs, which are no particular use right now, so they're also under the same namespace of `Location`.

I'm beginning to think you might want structured data for your Metadata! Or at least an optionally structured part to it, so we can add specific context to certain fields (like polls and locations!)

With post-processors that pull out data from messages I'd definitely want to do things like extract all the restaurants I've ever posted about (or visited!), or collect WebArchive copies of every website I posted about.

Even making a prettier UI for polls would mean agreement between front end and back on where the Poll metadata is stored. Agreement between two parts is an API; and an unstructured map as an API is asking for trouble!

What do you think? I'd probably put this in a `ReferencedLocation` _named_ key if this was in place.

mholt replied:
Thanks!

So, I'm currently thinking that the website docs will be for end users, but we can use the project wiki for developers. What do you think of that plan? Both will be collaborative.

jphastings replied:
Ahh yes of course — that sounds perfect; if you add a metadata-themed wiki page in the right place I'll add some context on these new keys!

mholt reviewed (approved):
Looking good now I think!

I did notice one final thing for now that is not really a showstopper. But before I finally merge this I wanted to get your thoughts.

mholt commented:
If you want to fix up the lint warnings (or add `nolint` comments if the reason for the magic number is obvious -- up to you), then I think we can merge this :)

mholt commented:
Nice, thank you for the contribution!
